### PR TITLE
[wip] Add RunAsGroup test in e2e PSP

### PR DESCRIPTION
**Pull request by mourya007**
Adding validateTokenRequest test file and missing test cases.
ref #73776 

What type of PR is this?
/kind cleanup

What this PR does / why we need it:
adds a Test 

Which issue(s) this PR fixes:

NONE

### CODE REVIEW

**Comment by mourya007** [comment]
cc @mikedanese

File: test/e2e/auth/pod_security_policy.go [hunk 1]
```diff
@@ -325,6 +329,9 @@ func restrictedPSP(name string) *policy.PodSecurityPolicy {
 			RunAsUser: policy.RunAsUserStrategyOptions{
 				Rule: policy.RunAsUserStrategyMustRunAsNonRoot,
 			},
+			RunAsGroup: &policy.RunAsGroupStrategyOptions{
+				Rule: policy.RunAsGroupStrategyMustRunAs,
+			},
```
Comment (krmayankk) [inline test/e2e/auth/pod_security_policy.go]: i think you would need to change the actual code that does the verification of  what the container is running as

Comment (mourya007) [reply test/e2e/auth/pod_security_policy.go]: I Could not understand what changes you want to change because It is already specified here https://github.com/kubernetes/kubernetes/blob/5647244b0c13db98816c136ad3e7d58551bbd41d/pkg/apis/policy/validation/validation.go#L236

**Comment by mourya007** [comment]
@liggitt  I have updated the comments for the remaining test cases.

File: test/e2e/auth/pod_security_policy.go [hunk 2]
```diff
@@ -270,6 +271,9 @@ func privilegedPSP(name string) *policy.PodSecurityPolicy {
 			RunAsUser: policy.RunAsUserStrategyOptions{
 				Rule: policy.RunAsUserStrategyRunAsAny,
 			},
+			RunAsGroup: &policy.RunAsGroupStrategyOptions{
+				Rule: policy.RunAsGroupStrategyMustRunAs,
```
Comment (krmayankk) [inline test/e2e/auth/pod_security_policy.go]: MustRunAs  should specify the id range the pod should run as and verify that pod in the range works without error and a pod outside that range should fail

Comment (mourya007) [reply test/e2e/auth/pod_security_policy.go]: Updated the comments.

**Comment by mourya007** [comment]
cc @mikedanese

File: test/e2e/auth/pod_security_policy.go (hunk 2 above)
Comment (krmayankk) [reply test/e2e/auth/pod_security_policy.go]: thanks , looks good, could you add one more test for negative case, when mustrunas is 999 and pod spec is 1000 , then create pod fails ?

File: pkg/apis/authentication/validation/validation_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		name string
+		obj  *authentication.TokenRequest
+		msg  string
+	}{
+		{
+			name: "specify larger time ",
+			obj: &authentication.TokenRequest{
+				ObjectMeta: metav1.ObjectMeta{Namespace: "a"},
+				Spec: authentication.TokenRequestSpec{
+					ExpirationSeconds: int64(7 * time.Second),
```
Comment (krmayankk) [inline pkg/apis/authentication/validation/validation_test.go]: i dont understand why this is a failure case ?

File: test/e2e/auth/pod_security_policy.go (hunk 1 above)
Comment (krmayankk) [reply test/e2e/auth/pod_security_policy.go]: my point was to verify whether the actual running container is indeed running as 999. I think we verify in other e2e. So here just verifying that PSP admission is working as expected is fine.

**Comment by krmayankk** [comment]
Adding @tallclair  for Selinux tests

**Comment by krmayankk** [comment]
also @mourya007 , as a small tip, its easier for reviewers to review, when you use new commits to address comments. when everything looks good, we squash them all in one.

File: pkg/apis/policy/validation/validation_test.go [hunk 4]
```diff
@@ -906,3 +906,30 @@ func TestValidatePSPRunAsGroup(t *testing.T) {
...
+func TestValidatePSPSELinux(t *testing.T) {
+	var testCases = []struct {
+		name    string
+		selinux policy.SELinuxStrategyOptions
+		fail    bool
+	}{
+		{"SELinuxStrategyMustRunAs",
+			policy.SELinuxStrategyOptions{
+				Rule:           policy.SELinuxStrategyMustRunAs,
+				SELinuxOptions: &api.SELinuxOptions{Level: "s9:z0,z1"}}, false},
```
Comment (tallclair) [inline pkg/apis/policy/validation/validation_test.go]: There should be at least one failure case.

File: pkg/apis/policy/validation/validation_test.go [hunk 5]
```diff
@@ -906,3 +906,30 @@ func TestValidatePSPRunAsGroup(t *testing.T) {
...
 	}
 }
+
+func TestValidatePSPSELinux(t *testing.T) {
+	var testCases = []struct {
+		name    string
+		selinux policy.SELinuxStrategyOptions
+		fail    bool
+	}{
+		{"SELinuxStrategyMustRunAs",
```
Comment (tallclair) [inline pkg/apis/policy/validation/validation_test.go]: nit: put the opening and closing braces on separate lines, to make the indentation more consistent.

File: pkg/apis/policy/validation/validation_test.go [hunk 6]
```diff
@@ -906,3 +906,30 @@ func TestValidatePSPRunAsGroup(t *testing.T) {
 		})
 	}
 }
+
+func TestValidatePSPSELinux(t *testing.T) {
```
Comment (tallclair) [inline pkg/apis/policy/validation/validation_test.go]: Thanks for adding the test, but it seems to be completely separate from the rest of this PR?

**Review by tallclair — COMMENTED** [review]
This PR appears to be adding 3 unrelated tests? Thank you for adding all the tests, but please separate them into 3 separate PRs.

File: pkg/apis/policy/validation/validation_test.go (hunk 6 above)
Comment (mourya007) [reply pkg/apis/policy/validation/validation_test.go]: Thanks for the review. Yes, I have added the test cases for different categories under the same PR. Can i separate them in 3 PR's?
Actually, I thought making 3 PR's just for the 3 test is seems more unfeasible.

File: test/e2e/auth/pod_security_policy.go (hunk 2 above)
Comment (liggitt) [reply test/e2e/auth/pod_security_policy.go]: agree we need a negative test that verifies a pod creation is rejected because of the runAsGroup field (ideally one testing runAsGroup in the spec.securityContext, and one testing runAsGroup in the spec.containers[0].securityContext)

File: pkg/apis/policy/validation/validation_test.go (hunk 6 above)
Comment (liggitt) [reply pkg/apis/policy/validation/validation_test.go]: the validation tests can be in a PR together. the e2e test should be separate.

the e2e test requires the RunAsGroup feature to be on, so it should be a distinct test tagged with `[Feature:RunAsGroup]` until https://github.com/kubernetes/kubernetes/pull/73007 merges

Comment (mourya007) [reply pkg/apis/policy/validation/validation_test.go]: Done, Sorry for the late as I was busy last week.

File: test/e2e/auth/pod_security_policy.go (hunk 2 above)
Comment (mourya007) [reply test/e2e/auth/pod_security_policy.go]: Sure, I will update the comments soon.

**Comment by krmayankk** [comment]
@mourya007 
- add RunAsGroup with id 65534 in restrictedPod()
- add  RunAsGroupStrategyMustRunAs with Id 65534 in restrictedPSP()
- add RunAsGroupStrategyRunAsAny in privilegedPSP()
- modify testPrivilegedPods to add two more cases, where you create a restrictedPod and modify the RunAsUser field to be 0 and RunAsGroup field of the pod to be 65529

and verify that all tests pass. Let me know if you face issues.

File: test/e2e/auth/pod_security_policy.go [hunk 7]
```diff
@@ -297,8 +309,9 @@ func restrictedPSP(name string) *policy.PodSecurityPolicy {
 			},
 		},
 		Spec: policy.PodSecurityPolicySpec{
-			Privileged:               false,
-			AllowPrivilegeEscalation: utilpointer.BoolPtr(false),
+			Privileged:                  false,
+			AllowPrivilegeEscalation:    utilpointer.BoolPtr(false),
+			RunAsGroupStrategyMustRunAs: utilpointer.Int64Ptr(65534),
```
Comment (krmayankk) [inline test/e2e/auth/pod_security_policy.go]: remove this

File: test/e2e/auth/pod_security_policy.go [hunk 8]
```diff
@@ -325,6 +338,11 @@ func restrictedPSP(name string) *policy.PodSecurityPolicy {
 			RunAsUser: policy.RunAsUserStrategyOptions{
 				Rule: policy.RunAsUserStrategyMustRunAsNonRoot,
 			},
+			RunAsGroup: &policy.RunAsGroupStrategyOptions{
+				Rule: policy.RunAsGroupStrategyMustRunAs,
+				Ranges: []policy.IDRange{
+					{Min: int64(999), Max: int64(999)}},
```
Comment (krmayankk) [inline test/e2e/auth/pod_security_policy.go]: make this id 65534

File: test/e2e/auth/pod_security_policy.go [hunk 9]
```diff
@@ -186,6 +186,12 @@ func testPrivilegedPods(tester func(pod *v1.Pod)) {
 		sysadmin.Spec.Containers[0].SecurityContext.AllowPrivilegeEscalation = nil
 		tester(sysadmin)
 	})
+
+	By("Running RunAsGroup and RunAsUser")
+	privileged := restrictedPod("RunAsGroup")
+	privileged.Spec.Containers[0].SecurityContext.RunAsGroup = 65529
+	privileged.Spec.Containers[0].SecurityContext.RunAsUser = 0
+	tester(privileged)
```
Comment (krmayankk) [inline test/e2e/auth/pod_security_policy.go]: it would be two different cases one with RunAsGroup as 65529 and another with RunAsUser as 0

Comment (krmayankk) [reply test/e2e/auth/pod_security_policy.go]: two different By() blocks

**Comment by mourya007** [comment]
I signed it

**Comment by mourya007** [comment]
"I signed it"

**Comment by Rajat-0** [comment]
Why we need the BUILD file here?

File: test/e2e/auth/pod_security_policy.go [hunk 10]
```diff
@@ -297,9 +316,9 @@ func restrictedPSP(name string) *policy.PodSecurityPolicy {
 			},
 		},
 		Spec: policy.PodSecurityPolicySpec{
-			Privileged:               false,
-			AllowPrivilegeEscalation: utilpointer.BoolPtr(false),
-			RequiredDropCapabilities: []v1.Capability{
+			Privileged:                  false,
+			AllowPrivilegeEscalation:    utilpointer.BoolPtr(false),
+		        RequiredDropCapabilities: []v1.Capability{
```
Comment (tallclair) [inline test/e2e/auth/pod_security_policy.go]: revert these changes

File: test/e2e/auth/pod_security_policy.go [hunk 11]
```diff
@@ -19,7 +19,7 @@ package auth
 import (
 	"fmt"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (tallclair) [inline test/e2e/auth/pod_security_policy.go]: revert this

File: pkg/apis/authentication/validation/BUILD [hunk 12]
```diff
@@ -1,26 +0,0 @@
...
-    tags = ["automanaged"],
-    visibility = ["//visibility:private"],
-)
-
-filegroup(
-    name = "all-srcs",
-    srcs = [":package-srcs"],
-    tags = ["automanaged"],
-    visibility = ["//visibility:public"],
-)
```
Comment (tallclair) [inline pkg/apis/authentication/validation/BUILD]: Revert this file

File: test/e2e/auth/pod_security_policy.go [hunk 13]
```diff
@@ -186,6 +186,19 @@ func testPrivilegedPods(tester func(pod *v1.Pod)) {
 		sysadmin.Spec.Containers[0].SecurityContext.AllowPrivilegeEscalation = nil
 		tester(sysadmin)
 	})
+	
+        By("Running RunAsGroup with id", func() {
+		privileged := restrictedPod("RunAsGroup")
+		privileged.Spec.Containers[0].SecurityContext.RunAsGroup = 65529
```
Comment (tallclair) [inline test/e2e/auth/pod_security_policy.go]: Use `0` instead, so it's more clear what this is doing.

**Comment by mourya007** [comment]
@tallclair Thanks for the review , just busy for this week. I will update the PR soon.

**Comment by krmayankk** [comment]
@mourya007 refer to this PR https://github.com/kubernetes/kubernetes/pull/75164 for fixing it . let us know whether you will be able to complete this in next couple days ?

**Comment by mourya007** [comment]
> @mourya007 refer to this PR #75164 for fixing it . let us know whether you will be able to complete this in next couple days ?

yes, I will but I think you have added the missing one in your PR. So what else is needed to fix here

**Comment by idealhack** [comment]
@krmayankk @mourya007 In case you don't know yet, you can use [`Co-authored-by`](https://help.github.com/en/articles/creating-a-commit-with-multiple-authors)

**Comment by mourya007** [comment]
@idealhack Yea, i have my individual CLA now, But it is Ok that mayank's PR has merged now. :)
